File: 0040_Combination Sum II/solution.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> combination;
        vector<int> path;
        dfs(candidates, target, path, combination, 0);
        return combination;
    }
    
    void dfs(vector<int> nums, int target, vector<int> path, vector<vector<int>>& combination, int index) {
        if (target == 0) {
            combination.push_back(path);
            return;
        }
        
        if (target < 0) {
            return;
        }
        
        for (int i = index; i < nums.size(); i++) {
            if (nums[i] > target) {
                break;
            }
            target += -nums[i];
            path.push_back(nums[i]);
            dfs(nums, target, path, combination, i + 1);
            target += nums[i];
            path.pop_back();
            
            while (i + 1 < nums.size() && nums[i + 1] == nums[i]) {
                i++;
            }
        }
    }
};

File: 0040_Combination Sum II/solution.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        combination = []  
        self.dfs(candidates, target, [], combination, 0)
        return combination
    
    def dfs(self, nums, target, path, combination, index):
        print(path)
        if target == 0:
            combination.append(path)
            return
        
        if target < 0:
            return
        
        for i in range(index, len(nums)):
            if i > index and nums[i] == nums[i - 1]:
                continue

            if nums[i] > target:
                break
                
            self.dfs(nums, target - nums[i], path + [nums[i]], combination, i + 1)